kqueen/engines/heat.py

```diff
@@ -0,0 +1,178 @@
+from collections import namedtuple
+import importlib
+import yaml
+import logging
+import sys
+import os
+
+from kqueen.config import current_config
+import openstack
+
+from .base import BaseEngine
+
+logger = logging.getLogger('kqueen_api')
+config = current_config()
+
+
+STATE_MAP = {
+    'IN_PROGRESS': config.get('CLUSTER_PROVISIONING_STATE'),
+    'COMPLETE': config.get('CLUSTER_OK_STATE'),
+}
+
+
+class Heat:
+    templates = {
+        "default": "kqueen.engines.resources.heat.default",
+        "empty": "kqueen.engines.resources.heat.empty",
+    }
+
+    def __init__(self, **kwargs):
+        self.connection = openstack.connection.Connection(**kwargs)
+
+    def _get_template_and_files(self, name):
+        # NOTE(sskripnick)
+        # we need to custom load and decode all files because of this bug:
+        # https://storyboard.openstack.org/#!/story/2002002
+        package = importlib.import_module(self.templates[name])
+        path = list(package.__path__)[0]
+        files = {}
+        for root, dirs, _files in os.walk(path):
+            for f in _files:
+                fpath = os.path.join(root, f)
+                files["file://" + fpath] = open(fpath, "rb").read().decode("utf8")
+        return os.path.join(path, "template.yaml"), files
```

Comment:
It is hard to support custom templates because we will need to support custom template parameters. We need to modify kqueen and kqueen-ui a lot to make this possible.

We also can support custom templates, but those ones should accept only predefined set of parameters, like `image_name`, `master_node_count`, `minion_count`, `ssh_key_name` etc.

Reply:
FYI, on UI we already have an ability to get a yaml/json files from user, validate them, and verify that required keys are provided